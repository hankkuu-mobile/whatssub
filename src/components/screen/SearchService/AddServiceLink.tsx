import * as React from 'react';

import styled from 'styled-components/native';

export interface AddServiceLinkProps {
  onClickAddServiceLink: () => void;
  style?: any;
}

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'salmon' : 'transparent')};
  height: 50;
  justify-content: center;
  align-items: center;
`;

const Touchable = styled.TouchableOpacity`
  background-color: ${({ theme }) => (theme.debug ? '#40E0D0' : 'transparent')};
  padding-top: 7;
  padding-bottom: 7;
  padding-left: 10;
  padding-right: 10;
`;

const StyledText = styled.Text`
  color: #407fff;
`;

function AddServiceLink({
  onClickAddServiceLink,
  ...rest
}: AddServiceLinkProps) {
  return (
    <Container {...rest}>
      <Touchable
        testID={'TOUCHABLE_ADD_SERVICE_LINK'}
        onPress={onClickAddServiceLink}
      >
        <StyledText>
          {'+ 검색 결과가 없다면 여기를 눌러 직접 입력해보세요.'}
        </StyledText>
      </Touchable>
    </Container>
  );
}

export default AddServiceLink;
