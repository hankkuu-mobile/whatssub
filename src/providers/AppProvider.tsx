import * as Font from 'expo-font';

import React, { useEffect, useReducer, useState } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [fontLoading, setFontLoading] = useState(true);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  const loadFonts = async () => {
    await Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Regular.ttf'),
    });

    setFontLoading(false);
  };
  useEffect(() => {
    loadFonts();
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {fontLoading && !props.doNotWaitFont ? null : props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
