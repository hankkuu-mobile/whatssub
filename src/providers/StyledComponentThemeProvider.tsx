import React, { useContext, useState } from 'react';

import { AppContext } from './AppProvider';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { createTheme } from '../theme';

const StyledComponentThemeProvider = (props) => {
  const { state } = useContext(AppContext);
  const [debug, setDebug] = useState(false);
  return (
    <ThemeProvider theme={{ ...createTheme(state.theme), debug, setDebug }}>
      {props.children}
    </ThemeProvider>
  );
};

StyledComponentThemeProvider.propTypes = {
  children: PropTypes.node,
};

export default StyledComponentThemeProvider;
