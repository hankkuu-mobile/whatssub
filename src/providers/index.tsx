import { AppConsumer, AppContext, AppProvider } from './AppProvider';

import PropTypes from 'prop-types';
import React from 'react';
import StyledComponentThemeProvider from './StyledComponentThemeProvider';
import { ThemeType } from '../types';

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

const InitialProviders = ({ children, ...rest }: Props) => (
  <AppProvider {...rest}>
    <StyledComponentThemeProvider>{children}</StyledComponentThemeProvider>
  </AppProvider>
);

InitialProviders.propTypes = {
  children: PropTypes.node,
};

export { AppProvider, AppConsumer, AppContext, InitialProviders };
