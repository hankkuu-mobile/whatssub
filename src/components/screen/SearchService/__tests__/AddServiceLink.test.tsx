import 'react-native';

import * as React from 'react';

import { fireEvent, render, wait } from '@testing-library/react-native';
import renderer, { act } from 'react-test-renderer';

import AddServiceLink from '../AddServiceLink';
import { InitialProviders } from '../../../../providers';

let component;
let props: any;

const createTestProps = (props: Object) => ({
  onClickAddServiceLink: jest.fn(),
});

const setup = () => {
  props = createTestProps({});
  component = (
    <InitialProviders doNotWaitFont>
      <AddServiceLink {...props} />
    </InitialProviders>
  );
};

describe('[AddServiceLink] render', () => {
  beforeAll(setup);
  it('component and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(component)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[AddServiceLink] interaction', () => {
  beforeAll(setup);
  it('should call onClickAddServiceLink', () => {
    const { getByTestId } = render(component);
    const AddServiceLink = getByTestId('TOUCHABLE_ADD_SERVICE_LINK');
    act(() => {
      fireEvent.press(AddServiceLink);
    });
    expect(props.onClickAddServiceLink).toBeCalled();
  });
});
