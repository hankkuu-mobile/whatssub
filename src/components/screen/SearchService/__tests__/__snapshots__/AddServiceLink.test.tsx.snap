// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AddServiceLink] render component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "height": 50,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "opacity": 1,
        "paddingBottom": 7,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 7,
      }
    }
    testID="TOUCHABLE_ADD_SERVICE_LINK"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#407fff",
          },
        ]
      }
    >
      + 검색 결과가 없다면 여기를 눌러 직접 입력해보세요.
    </Text>
  </View>
</View>
`;
